x-pack/apm-server/aggregation/servicesummarymetrics/aggregator.go

```diff
@@ -177,44 +200,61 @@ func newMetricsBuffer(maxSize int) *metricsBuffer {
 	return &metricsBuffer{
 		maxSize: maxSize,
 		m:       make(map[uint64][]*aggregationKey),
+
+		// Keep one reserved entry for overflow bucket.
+		space: make([]aggregationKey, maxSize+1),
 	}
 }
 
 func (mb *metricsBuffer) storeOrUpdate(
 	key aggregationKey,
+	interval time.Duration,
 	logger *logp.Logger,
-) bool {
+) {
 	hash := key.hash()
 
 	mb.mu.Lock()
 	defer mb.mu.Unlock()
 
-	entries, ok := mb.m[hash]
-	if ok {
-		ok = false
-		for _, old := range entries {
-			if old.equal(key) {
-				ok = true
-				break
+	find := func(hash uint64, key aggregationKey) (*aggregationKey, bool) {
+		// This function should only be called when caller is holding the lock mb.mu.
+		// It takes separate hash and key arguments so that hash can be computed
+		// before acquiring the lock.
+		entries, ok := mb.m[hash]
+		if ok {
+			for _, old := range entries {
+				if old.equal(key) {
+					return old, true
+				}
 			}
 		}
+		return nil, false
+	}
+
+	_, ok := find(hash, key)
+	if ok {
+		return
 	}
 
-	if !ok {
-		n := len(mb.m)
-		half := mb.maxSize / 2
-
-		switch n {
-		case mb.maxSize:
-			return false
-		case half - 1:
-			logger.Warn("service summary groups reached 50% capacity")
-		case mb.maxSize - 1:
-			logger.Warn("service summary groups reached 100% capacity")
+	if mb.entries >= mb.maxSize {
+		if mb.otherCardinalityEstimator == nil {
+			logger.Warnf(`
+Service summary aggregation group limit of %d reached, new metric documents will be grouped
+under a dedicated bucket identified by service name '%s'.`[1:], mb.maxSize, overflowServiceName)
+			mb.space[len(mb.space)-1] = makeOverflowAggregationKey(interval)
+			mb.other = &mb.space[len(mb.space)-1]
+			mb.otherCardinalityEstimator = hyperloglog.New14()
 		}
-		mb.m[hash] = append(mb.m[hash], &key)
+		mb.otherCardinalityEstimator.InsertHash(hash)
+		return
+	} else if mb.entries == mb.maxSize/2-1 {
+		logger.Warn("service summary groups reached 50% capacity")
+	} else if mb.entries == mb.maxSize-1 {
+		logger.Warn("service summary groups reached 100% capacity")
```

Comment:
We are not issuing these warnings anymore for the service metrics or the transaction metrics. Can you align the overflow logic where possible with these aggregations?